docker-compose.yml

```diff
@@ -16,16 +16,22 @@ services:
       READ_ONLY_API: ${READ_ONLY_API:-}
       VNC_SERVER_PASSWORD: ${VNC_SERVER_PASSWORD:-}
       TWOFA_TIMEOUT_ACTION: ${TWOFA_TIMEOUT_ACTION:-exit}
+      BYPASS_WARNING: ${BYPASS_WARNING:-}
       AUTO_RESTART_TIME: ${AUTO_RESTART_TIME:-}
+      AUTO_LOGOFF_TIME: ${AUTO_LOGOFF_TIME:-}
+      SAVE_TWS_SETTINGS: ${SAVE_TWS_SETTINGS:-}
       RELOGIN_AFTER_TWOFA_TIMEOUT: ${RELOGIN_AFTER_TWOFA_TIMEOUT:-no}
       TWOFA_EXIT_INTERVAL: ${TWOFA_EXIT_INTERVAL:-60}
       TIME_ZONE: ${TIME_ZONE:-Etc/UTC}
+      GATEWAY_OR_TWS: ${GATEWAY_OR_TWS:--g}
       CUSTOM_CONFIG: ${CUSTOM_CONFIG:-NO}
+      SSH_PUBLIC_KEY: ${SSH_PUBLIC_KEY}
```

Comment:
I don't see many options to protect a container in the cloud running TWS with api sockets open on a public internet network, than shielding these ports (as well as VNC port) with SSH. The client will use its private key to authenticat to the server running TWS (or Gateway).
Another tool to protect it, could be fail2Ban

Reply:
there is already SSH support on the container. see --> https://github.com/gnzsnz/ib-gateway-docker#ssh-tunnel

SSH support is implemented with ssh-client, which means that the container can connect an ssh server. not the other way around. this avoid implementing an ssh-server,  which is an anti-patern. please remove all changes related to ssh server.

this one is a "no"

Reply:
OK. I'm creating a local branch just for me keeping ssh server because in my flow, it's the container running the trading bot logic which connects to the instance running TWS.

Reply:
Hi, you can implement the [tunnel](https://github.com/gnzsnz/ib-gateway-docker#ssh-tunnel) with minimal changes on your current logic.

the idea behind containers is that they should do one thing, an SSH server requires setup, maintenance and is a risk. that's why it should be in a separate container, I have build a container for [SSH](https://github.com/gnzsnz/docker-bastion) to address this. there are other containers available, so there is choice.

based on your case, you can 

1. set-up an SSH container(most of the work is to setup user/host Keys, using [SSH certificates][1] simplifies this.)
2. connect your trading bot to the SSH container, this is a minimal change. just change the host you connect to.
3. connect ib-gateway to SSH container, this is already supported

your ib-gatewa is secured, your trading bot is secured, and you use a security hardened SSH container. Tunnels eventually get disconnected, but is the same than with your current setup. ib-gateway has a logic to restart tunnels so this is mitigated.

i'm happy to help on this one.

for reference: https://goteleport.com/blog/how-to-ssh-properly/


[1]: https://gist.github.com/gnzsnz/3fcfa368fa3bc181c8bd69da96037b2c